File: web-app/src/services/notificationWebSocketService.ts
import { notificationStore } from "@/utils/notificationStore";
import type { NotificationWebSocketMessage } from "@/types/notification";
import { playSound } from "@/utils/soundUtils";

class NotificationWebSocketService {
  private ws: WebSocket | null = null;
  private reconnectTimeout: NodeJS.Timeout | null = null;
  private pingInterval: NodeJS.Timeout | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private token: string | null = null;
  private isConnecting = false;

  constructor() {
    this.connect = this.connect.bind(this);
    this.disconnect = this.disconnect.bind(this);
    this.send = this.send.bind(this);
  }

  connect(token: string): void {
    if (
      this.isConnecting ||
      (this.ws && this.ws.readyState === WebSocket.OPEN)
    ) {
      return;
    }

    this.token = token;
    this.isConnecting = true;
    notificationStore.setError(null);

    try {
      // Always derive protocol from API URL to ensure consistency
      const apiUrl =
        process.env.NEXT_PUBLIC_API_URL || "http://localhost:8080/api/v1";
      const apiBase = apiUrl.replace(/\/api\/v1$/, ""); // Remove /api/v1 suffix
      const wsProtocol = apiBase.startsWith("https") ? "wss" : "ws";

      // If NEXT_PUBLIC_WS_URL is set, extract host/port from it, otherwise use API URL
      let hostAndPath = apiBase.replace(/^https?:\/\//, ""); // Remove http:// or https://
      if (process.env.NEXT_PUBLIC_WS_URL) {
        // Extract host/port from WS URL (ignore protocol)
        const wsUrlMatch =
          process.env.NEXT_PUBLIC_WS_URL.match(/^wss?:\/\/(.+)$/);
        if (wsUrlMatch) {
          hostAndPath = wsUrlMatch[1];
        }
      }

      const baseUrl = `${wsProtocol}://${hostAndPath}`;
      const wsUrl = `${baseUrl}/api/v1/in-app-notifications/ws?token=${token}`;

      this.ws = new WebSocket(wsUrl);

      this.ws.onopen = () => {
        this.isConnecting = false;
        this.reconnectAttempts = 0;
        this.startPing();
        notificationStore.setError(null);
      };

      this.ws.onmessage = (event) => {
        try {
          const message: NotificationWebSocketMessage = JSON.parse(event.data);
          this.handleMessage(message);
        } catch {
          // Ignore parsing errors
        }
      };

      this.ws.onclose = (event) => {
        this.isConnecting = false;
        this.ws = null;
        this.stopPing();

        // Attempt to reconnect if not a normal closure
        if (event.code !== 1000 && this.token) {
          this.scheduleReconnect();
        }
      };

      this.ws.onerror = () => {
        this.isConnecting = false;
        notificationStore.setError("WebSocket connection error");
      };
    } catch {
      this.isConnecting = false;
      notificationStore.setError("Failed to connect to WebSocket");
    }
  }

  disconnect(): void {
    if (this.reconnectTimeout) {
      clearTimeout(this.reconnectTimeout);
      this.reconnectTimeout = null;
    }

    this.stopPing();

    if (this.ws) {
      this.ws.close(1000, "Client disconnecting");
      this.ws = null;
    }

    this.token = null;
    this.reconnectAttempts = 0;
  }

  send(message: Record<string, unknown>): void {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(message));
    } else {
    }
  }

  private handleMessage(message: NotificationWebSocketMessage): void {
    switch (message.event) {
      case "new_notification":
        if (message.data.notification) {
          notificationStore.addNotification(message.data.notification);
          // Play notification sound for new notifications
          playSound("notification");
        }
        if (message.data.unread_count !== undefined) {
          notificationStore.setUnreadCount(message.data.unread_count);
        }
        break;

      case "unread_count_update":
        if (message.data.unread_count !== undefined) {
          notificationStore.setUnreadCount(message.data.unread_count);
        }
        break;

      case "notification_read":
        if (message.data.notification_id) {
          notificationStore.markAsRead(message.data.notification_id);
        }
        break;

      case "all_notifications_read":
        notificationStore.markAllAsRead();
        break;

      case "notification_updated":
        if (message.data.notification) {
          notificationStore.updateNotification(
            message.data.notification.id,
            message.data.notification
          );
        }
        break;

      default:
    }
  }

  private scheduleReconnect(): void {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      notificationStore.setError("Connection lost. Please refresh the page.");
      return;
    }

    const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempts), 30000);
    this.reconnectAttempts++;

    this.reconnectTimeout = setTimeout(() => {
      if (this.token) {
        this.connect(this.token);
      }
    }, delay);
  }

  private startPing(): void {
    this.pingInterval = setInterval(() => {
      if (this.ws && this.ws.readyState === WebSocket.OPEN) {
        this.send({ type: "ping" });
      }
    }, 30000); // Ping every 30 seconds
  }

  private stopPing(): void {
    if (this.pingInterval) {
      clearInterval(this.pingInterval);
      this.pingInterval = null;
    }
  }

  get isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }
}

// Create and export singleton instance
export const notificationWebSocketService = new NotificationWebSocketService();

File: flutter/lib/pages/notifications_page/data/services/notification_websocket_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:developer';
import 'package:trees_india/commons/environment/global_environment.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

enum WebSocketConnectionState {
  disconnected,
  connecting,
  connected,
  error,
}

class NotificationWebSocketMessage {
  final String event;
  final Map<String, dynamic> data;

  NotificationWebSocketMessage({
    required this.event,
    required this.data,
  });

  factory NotificationWebSocketMessage.fromJson(Map<String, dynamic> json) {
    return NotificationWebSocketMessage(
      event: json['event'] as String,
      data: json['data'] as Map<String, dynamic>,
    );
  }
}

class NotificationWebSocketService {
  WebSocketChannel? _channel;
  final StreamController<NotificationWebSocketMessage> _messageController =
      StreamController<NotificationWebSocketMessage>.broadcast();
  final StreamController<WebSocketConnectionState> _connectionStateController =
      StreamController<WebSocketConnectionState>.broadcast();
  final StreamController<String?> _errorController =
      StreamController<String?>.broadcast();

  Timer? _reconnectTimer;
  Timer? _pingTimer;
  int _reconnectAttempts = 0;
  static const int _maxReconnectAttempts = 5;
  String? _token;
  bool _isConnecting = false;

  // Streams
  Stream<NotificationWebSocketMessage> get messageStream =>
      _messageController.stream;
  Stream<WebSocketConnectionState> get connectionStateStream =>
      _connectionStateController.stream;
  Stream<String?> get errorStream => _errorController.stream;

  // Current state
  WebSocketConnectionState get connectionState =>
      _connectionStateController.hasListener
          ? WebSocketConnectionState.disconnected
          : WebSocketConnectionState.disconnected;

  bool get isConnected => _channel != null;

  void connect(String token) {
    if (_isConnecting || (_channel != null && isConnected)) {
      return;
    }

    _token = token;
    _isConnecting = true;
    _connectionStateController.add(WebSocketConnectionState.connecting);
    _errorController.add(null);

    try {
      final wsBaseUrl = GlobalEnvironment.wsBaseUrl;
      final wsUrl = '$wsBaseUrl/api/v1/in-app-notifications/ws';
      final uri = Uri.parse('$wsUrl?token=$token');

      _channel = WebSocketChannel.connect(uri);

      _channel!.ready.then((_) {
        _isConnecting = false;
        _reconnectAttempts = 0;
        _startPing();
        _connectionStateController.add(WebSocketConnectionState.connected);
        log('🔌 WebSocket connected successfully');
      }).catchError((error) {
        _isConnecting = false;
        _handleConnectionError(error);
      });

      _channel!.stream.listen(
        (message) {
          try {
            final Map<String, dynamic> data = json.decode(message);
            final wsMessage = NotificationWebSocketMessage.fromJson(data);
            _messageController.add(wsMessage);
            log('📩 Received WebSocket message: ${wsMessage.event}');
          } catch (e) {
            log('❌ Error parsing WebSocket message: $e');
          }
        },
        onError: (error) {
          log('❌ WebSocket stream error: $error');
          _handleConnectionError(error);
        },
        onDone: () {
          log('🔌 WebSocket connection closed');
          _handleConnectionClosed();
        },
      );
    } catch (e) {
      _isConnecting = false;
      _handleConnectionError(e);
    }
  }

  void disconnect() {
    log('🔌 Disconnecting WebSocket');
    _stopReconnectTimer();
    _stopPing();

    _channel?.sink.close(1000, 'Client disconnecting');
    _channel = null;
    _token = null;
    _reconnectAttempts = 0;
    _connectionStateController.add(WebSocketConnectionState.disconnected);
  }

  void send(Map<String, dynamic> message) {
    if (_channel != null && isConnected) {
      _channel!.sink.add(json.encode(message));
      log('📤 Sent WebSocket message: ${message['type'] ?? 'unknown'}');
    } else {
      log('❌ Cannot send message: WebSocket not connected');
    }
  }

  void markAsRead(int notificationId) {
    send({
      'type': 'mark_read',
      'notification_id': notificationId,
    });
  }

  void markAllAsRead() {
    send({
      'type': 'mark_all_read',
    });
  }

  void _handleConnectionError(dynamic error) {
    log('❌ WebSocket connection error: $error');
    _isConnecting = false;
    _channel = null;
    _stopPing();
    _connectionStateController.add(WebSocketConnectionState.error);
    _errorController.add(error.toString());
    _scheduleReconnect();
  }

  void _handleConnectionClosed() {
    _isConnecting = false;
    _channel = null;
    _stopPing();
    _connectionStateController.add(WebSocketConnectionState.disconnected);

    if (_token != null) {
      _scheduleReconnect();
    }
  }

  void _scheduleReconnect() {
    if (_reconnectAttempts >= _maxReconnectAttempts) {
      _errorController.add('Connection lost. Please refresh the page.');
      return;
    }

    final delay = Duration(
      milliseconds: (1000 * (1 << _reconnectAttempts)).clamp(1000, 30000),
    );
    _reconnectAttempts++;

    _reconnectTimer = Timer(delay, () {
      if (_token != null) {
        log('🔄 Attempting WebSocket reconnect (attempt $_reconnectAttempts)');
        connect(_token!);
      }
    });
  }

  void _startPing() {
    _pingTimer = Timer.periodic(const Duration(seconds: 30), (timer) {
      if (_channel != null && isConnected) {
        send({'type': 'ping'});
      }
    });
  }

  void _stopPing() {
    _pingTimer?.cancel();
    _pingTimer = null;
  }

  void _stopReconnectTimer() {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
  }

  void dispose() {
    disconnect();
    _messageController.close();
    _connectionStateController.close();
    _errorController.close();
  }
}
